### PR TITLE
Refactor Node.nodes to contain both components and node links

### PR DESCRIPTION
**Pull request by sloria**
## Changes

* The `AbstractNode.nodes` field is now a M2M with `AbstractNode` through a new `NodeRelation` model
* This allows both subprojects and node links to be stored under the same field, and thus ordered amongst each other

## TODO

- [x] Make `AbstractNode.nodes.all()` respect ordering 
- [x] Rename `source` and `dest` to `parent` and `child` for clarity?
- [x] Update osf_models_tests
- [ ] Update v1 code and tests/test_views.py
- [x] Update apiv2 code and api_tests
- [ ] migration code

### CODE REVIEW

File: osf_models/models/node.py [hunk 1]
```diff
@@ -1316,28 +1337,31 @@ def register_node(self, schema, auth, data, parent=None):
...
-                data=data,
-                parent=registered,
-            )
-            child_pks.append(registered_child.pk)
-        # Preserve ordering of children
-        registered.set_abstractnode_order(child_pks)
-
-        # Copy linked nodes
-        registered.linked_nodes.add(*original.linked_nodes.values_list('pk', flat=True))
+        for node_relation in original.node_relations.filter(dest__is_deleted=False):
```
Comment (cwisecarver) [inline osf_models/models/node.py]: You might be able to use something [here](http://stackoverflow.com/questions/3699395/find-parent-recursively-using-query#6431757) do this all at the database.

I used the third or forth example to build the materialized path of a file in one query.
https://github.com/CenterForOpenScience/osf-models/blob/master/osf_models/models/files.py#L68

Comment (cwisecarver) [inline osf_models/models/node.py]: This comment was actually supposed to reference https://github.com/CenterForOpenScience/osf-models/pull/141/files#diff-4c72132379ea99ca4dd6dbac66040560L1384

Comment (sloria) [inline osf_models/models/node.py]: Ooh, very cool. I might hold off on that for now, but looks like a viable solution.

File: osf_models/models/node_relation.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+from .base import BaseModel, ObjectIDMixin
+
+class NodeRelation(ObjectIDMixin, BaseModel):
+
+    source = models.ForeignKey('AbstractNode', related_name='node_relations')
+    dest = models.ForeignKey('AbstractNode')
+    is_node_link = models.BooleanField(default=False, db_index=True)
+
+    class Meta:
+        order_with_respect_to = 'source'
```
Comment (sloria) [inline osf_models/models/node_relation.py]: This is the bit that's stumping me at the moment. I can't figure out a way to access `Node.nodes.all()` such that they are ordered by `NodeRelation._order`. I've tried:

- `node.nodes_order_by('node_relations')`
- `node.nodes.order_by('node_relations___order')`
- setting `related_name='node_relations_dest` on `dest` and doing  `node.nodes.order_by('node_relations_dest___order')`

But none of these are working.

Comment (cwisecarver) [inline osf_models/models/node_relation.py]: Okay... so... `order_with_respect_to` doesn't officially work with m2m fields...

Best I could do (without rewriting m2m field and manager) was 

```python
[nr.dest for nr in NodeRelation.objects.filter(source=n).select_related('dest')]
```

It does one query to get the relations and their nodes and then does one query each for the guids (which shouldn't happen in the new world order).

**Comment by sloria** [comment]
OK, this should mostly be working, using the approach discussed with @cwisecarver . 

Next step is to get the osf code updated and the models tests passing.

**Comment by sloria** [comment]
There are still failing views tests, but these are also failing on master. In the interest of getting to staging3, going to go ahead and merge this.

APIv2 and a good chunk of v1 code has been updated to work with this change.

**Pull request merged**